src/tools/rbd_mirror/image_map/types.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <set>
+#include <map>
+#include <string>
+#include <memory>
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MapShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct MappedImage {
+  std::string id;
+  std::string global_id;
+  std::string instance_id;
+
+  MappedImage(const std::string &global_id)
+    : global_id(global_id) {
+  }
+
+  MappedImage(const std::string &id,
+              const std::string &global_id)
+    : id(id), global_id(global_id) {
+  }
+
+  MappedImage(
+    const std::string &id,
+    const std::string &global_id,
+    const std::string &instance_id)
+    : id(id),
+      global_id(global_id),
+      instance_id(instance_id) {
+  }
+
+  inline bool operator==(const MappedImage &rhs) const {
+    return (global_id == rhs.global_id && id == rhs.id);
+  }
+
+  inline bool operator<(const MappedImage &rhs) const {
+    return global_id < rhs.global_id;
+  }
+};
+
+struct RemappedImage {
```

Comment:
These structures are used to pass a bunch of images to `PoolReplayer::handle_update()`. Tracking next instance id (plus other things) should mostly be a part of some other "long lived" data structure such as the map itself (or a part of the map which tracks peer uuids).

Regarding the `RemappedImage` structure, yeh, we have just stick to having just one structure for representing mapped, removed and remapped images.